cmd/veneur-prometheus/main.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/Sirupsen/logrus"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+var (
+	debug       = flag.Bool("d", false, "Enable debug mode")
+	metricsHost = flag.String("h", "http://localhost:9090/metrics", "The full URL — like 'http://localhost:9090/metrics' to query for Prometheus metrics.")
+	interval    = flag.String("i", "10s", "The interval at which to query. Value must be parseable by time.ParseDuration (https://golang.org/pkg/time/#ParseDuration).")
+	statsHost   = flag.String("s", "127.0.0.1:8126", "The host and port — like '127.0.0.1:8126' — to send our metrics to.")
+)
+
+func main() {
+	flag.Parse()
+
+	c, _ := statsd.New(*statsHost)
+
+	if *debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+	logrus.Debug("HELLO")
+
+	i, err := time.ParseDuration(*interval)
+	if err != nil {
+		logrus.WithError(err).Fatalf("Failed to parse interval '%s'", *interval)
+	}
+
+	ticker := time.NewTicker(i)
+	for _ = range ticker.C {
+		collect(c)
+	}
+}
+
+func collect(c *statsd.Client) {
+	logrus.WithFields(logrus.Fields{
+		"stats_host":   *statsHost,
+		"metrics_host": *metricsHost,
+	}).Debug("Beginning collection")
+
+	resp, _ := http.Get(*metricsHost)
+	d := expfmt.NewDecoder(resp.Body, expfmt.FmtText)
+	var mf dto.MetricFamily
+	for {
+		err := d.Decode(&mf)
```

Comment:
I'm assuming this decoder is guaranteed to be exact? Otherwise we're going to get a lot of problems. I don't see a `ReadByte` method either.

Reply:
Sorry, I'm not sure what you mean here. :( Exact?

Reply:
Basically, can `Decode` ever read more data than it needs (possibly into some internal buffer). The `zlib` reader does that, which I learned the hard way - the presence of a `ReadByte` method is usually the way to "force" exact behavior, though that depends on the reader honoring this contract (it's not enforced).

Since we're reusing the same reader here between each iteration of the loop, it's unlikely that the `Decode` would read (and discard) extra data, as opposed to possibly keeping that in an internal buffer that's used when `Decode` is called again in the next loop. But I figured I'd ask.

Given our current uses of prometheus (non-critical), I'm fine with confirming this by testing it out, but we should then emit a metric to track decoding failures (a huge spike in that error will indicate that the decoder is *not* exact).

Reply:
Gotya! We talked IRL but this has been revamped with detection of `io.EOF` and counters for other errors. I think we're good!